### ISSUE TITLE
Unable to call apply on exported Procedure value

### ISSUE THREAD

**Issue by MichaelBurge** [racket/typed-racket#668]
A Procedure cannot be called in Typed Racket because nothing is known about it. When I have a collection of functions of varying arities, I will make their type Procedure and use a separate "unsafe" module with "#lang racket" or "#lang typed/racket/no-check" to call them.

However, it seems it's currently impossible to `apply` Procedure values even in an unsafe module. The values must both be created and invoked in an unsafe module.

### What version of Racket are you using?

6.12

### What program did you run?

```scheme
#lang racket

(module a typed/racket
  (provide a)

  (: a Procedure)
  (define (a x) x)
  )

(require 'a)

(procedure-arity a)
(apply a '(5))
```

### What should have happened?

The program should print "1" and "5". Instead, it prints "1" and then throws an error.

### If you got an error message, please include it here.
```
1
g5: arity mismatch;
 the expected number of arguments does not match the given number
  given: 1
  arguments...:
   5
  context...:
   /home/mburge/work/tmp/a.rkt: [running body]
```

The error is especially confusing because (procedure-arity) is 1, the error message mentions the function is given 1 argument, yet the error mentions the number of arguments is incorrect.

The `g5` function appears to take 0 arguments, but invokes the `a` function which requires 1 argument. So there is no way to invoke `a` since the same argument list is used for both `g5` and `a`.

**Comment by samth** [comment]
An untyped module isn't "unsafe", it's still protected by a contract. There's no way to safely call a value of type `Procedure`, either in typed or untyped code -- `apply` doesn't let you get around that.

**Comment by samth** [comment]
The error message and `procedure-arity` result are the same as in this program, which uses the same contract that Typed Racket is generating:

```
#lang racket
(define/contract f (case->) (λ (x) x))
(procedure-arity f)
(apply f '(1))
```

**Comment by MichaelBurge** [comment]
Perhaps the contract could give a better error message like "Attempted to call a value of type `Procedure`."

Is the issue that Typed Racket needs dependent types (using `procedure-arity` or `procedure-arity-includes?`) to be able to express when it's safe to call a `Procedure`?

Ideally, this would be safe:
```scheme
(: proc Procedure )
(define (proc x) x)
(define arity (procedure-arity proc))
(define args '(5))

(if (procedure-arity-includes? proc (length args))
  (apply proc args)
  (error "Incorrect number of arguments" arity args))
```

**Comment by AlexKnauth** [comment]
Even then you have to consider that the “original” typed function might require a restrictions on it’s argument types, even when the arity is correct.
```racket
(provide public-version)

(: orig : Nothing -> Integer)
(define (orig x) x)

(: public-version : Procedure)
(define public-version orig)
```

Applying `public-version` or `orig` to anything should be an error.

**Comment by samth** [comment]
Yes, as @AlexKnauth points out, there's no way to tell dynamically what the "safe" arguments to a function are, and thus there's no way to ever safely call something of type `Procedure`, even with a fancier type system.

As to the error message, @rfindler could the contract error message for `(case->)` change, or is that something that would have a performance impact?

**Comment by rfindler** [comment]
I don't see a program that uses (only) contracts.

**Comment by samth** [comment]
@rfindler this comment has such a program: https://github.com/racket/typed-racket/issues/668#issuecomment-368070289

**Comment by mfelleisen** [comment]
@samth I disagree. I think an occurrence type system could lower the type of a Procedure value p to (Any ->* Any) with (if (is-arity-of p 1) ... 42). TR's doesn't do so but I wouldn't say it's impossible.

**Comment by samth** [comment]
@mfelleisen no, that's not safe -- see the example @AlexKnauth posted. You could lower it to `(-> Nothing Any)`, but you still can't call that.

**Comment by rfindler** [comment]
Just to be clear, the problem with the program that @samth posted:

```
#lang racket
(define/contract f (case->) (λ (x) x))
(procedure-arity f)
(apply f '(1))
```

is not the error message, but the result of `procedure-arity`, which should have been an empty list?

In other words, this program is fine?

```
#lang racket
(define f (case-lambda))
(procedure-arity f)
(f 1)
```

**Comment by samth** [comment]
@rfindler That's what I think seems reasonable. Maybe @MichaelBurge wants something different, though.

**Comment by rfindler** [comment]
Looking a bit more, I'm not sure that this is something that I can fix with changes that are only at the contract system layer. It happens with arrow, too:

```
> (procedure-arity (contract (-> integer? integer?) (λ x x) 'pos 'neg))
(arity-at-least 0)
```

and more generally if I try to reduce the arity of a function, it isn't `chaperone-of` anymore:

```
> (chaperone-of?
   (procedure-reduce-arity list 1)
   list)
#f
```
